File: LeetCode_Monthly_Challenge/July/Java/Week_3/Remove Linked List Elements.java
/*
PROBLEM:
Remove all elements from a linked list of integers that have value val.

Example:
Input:  1->2->6->3->4->5->6, val = 6
Output: 1->2->3->4->5
*/



/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
 
 
 
 
class Solution {
    public ListNode removeElements(ListNode head, int val) {
        if(head == null) return head;
        ListNode prev = null, curr = head;
        while(curr != null){
            if(curr.val == val)
                if(prev == null) head = curr.next;
                else prev.next = curr.next;
            else
                prev = curr;
            curr = curr.next;
        }
        return head;
    }
}

File: LeetCode_Monthly_Challenge/July/Python3/Week_3/Remove Linked List Elements.py
""" 
PROBLEM:
Remove all elements from a linked list of integers that have value val.

Example:
Input:  1->2->6->3->4->5->6, val = 6
Output: 1->2->3->4->5
"""



# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next



class Solution:
    def removeElements(self, head: ListNode, val: int) -> ListNode:
        if head is None: return head
        prev, curr = None, head
        while curr:
            if curr.val == val:
                if prev is None: head = curr.next
                else: prev.next = curr.next
            else:
                prev = curr
            curr = curr.next
        return head
